File: frontend/src/store/modules/notification.ts
/** 状态管理模块：notification。 */
import { defineStore } from 'pinia'
import {
  fetchInAppNoticeList,
  fetchReadAllInAppNotice,
  fetchReadInAppNotice,
  type InAppNoticePage
} from '@/api/notifications'
import { useUserStore } from './user'

type NotificationWsEnvelope = {
  type: string
  version: 1
  sequence: number
  occurredAt: string
  data: Record<string, any>
}

export function decodeNotificationWsEnvelope(
  value: unknown,
  lastSequence: number
): NotificationWsEnvelope | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return null
  }
  const envelope = value as Record<string, unknown>
  const sequence = typeof envelope.sequence === 'number' ? envelope.sequence : Number.NaN
  if (
    envelope.version !== 1 ||
    typeof envelope.type !== 'string' ||
    envelope.type === '' ||
    !Number.isSafeInteger(sequence) ||
    sequence <= lastSequence ||
    typeof envelope.occurredAt !== 'string' ||
    !envelope.occurredAt.endsWith('Z') ||
    !envelope.data ||
    typeof envelope.data !== 'object' ||
    Array.isArray(envelope.data)
  ) {
    return null
  }
  return { ...envelope, version: 1, sequence } as NotificationWsEnvelope
}

const NOTIFICATION_WS_PROTOCOL = 'coinsphere.notifications.v1'

export function buildNotificationWsUrl(pageOrigin: string) {
  const url = new URL('/api/v1/ws/notifications', pageOrigin)
  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
    throw new Error('notification websocket requires an HTTP page origin')
  }
  url.protocol = url.protocol === 'https:' ? 'wss:' : 'ws:'
  return url.toString()
}

export const useNotificationStore = defineStore('notificationStore', () => {
  const records = ref<Api.Notifications.InAppNoticeItem[]>([])
  const unreadCount = ref(0)
  const size = ref(20)
  const nextCursor = ref('')
  const total = ref(0)
  const hasMore = ref(false)
  const loading = ref(false)
  const connected = ref(false)
  const latestIncomingNoticeId = ref<number | null>(null)

  let socket: WebSocket | null = null
  let reconnectTimer: ReturnType<typeof setTimeout> | null = null
  let manualClose = false
  let lastSequence = 0

  const userStore = useUserStore()

  const resetState = () => {
    records.value = []
    unreadCount.value = 0
    nextCursor.value = ''
    total.value = 0
    hasMore.value = false
    loading.value = false
    latestIncomingNoticeId.value = null
  }

  const applyPage = (page: InAppNoticePage, append = false) => {
    const normalizedRecords = page.records.map(normalizeNoticeRecord)
    records.value = append ? [...records.value, ...normalizedRecords] : normalizedRecords
    unreadCount.value = page.unreadCount
    nextCursor.value = page.nextCursor
    total.value = page.total
    hasMore.value = page.hasMore
  }

  const loadNotices = async (options?: { append?: boolean }) => {
    if (loading.value || userStore.accessMode !== 'authenticated') {
      return
    }
    loading.value = true
    try {
      const page = await fetchInAppNoticeList({
        cursor: options?.append ? nextCursor.value || undefined : undefined,
        limit: size.value
      })
      applyPage(page, Boolean(options?.append))
    } finally {
      loading.value = false
    }
  }

  const markRead = async (deliveryId: number) => {
    await fetchReadInAppNotice(deliveryId)
    const record = records.value.find((item) => item.id === deliveryId)
    if (record && !record.isRead) {
      record.isRead = true
      record.readAt = new Date().toISOString()
      unreadCount.value = Math.max(unreadCount.value - 1, 0)
    }
  }

  const markAllRead = async () => {
    await fetchReadAllInAppNotice()
    records.value = records.value.map((item) => ({
      ...item,
      isRead: true,
      readAt: item.readAt || new Date().toISOString()
    }))
    unreadCount.value = 0
  }

  const applySignalDecision = (signalId: string, status: string) => {
    records.value = records.value.map((item) =>
      item.strategySignalId === signalId ? { ...item, strategySignalStatus: status } : item
    )
  }

  const handleWsMessage = (envelope: NotificationWsEnvelope) => {
    if (envelope.type === 'notice.unread') {
      const nextUnreadCount = envelope.data.unreadCount
      if (
        typeof nextUnreadCount === 'number' &&
        Number.isSafeInteger(nextUnreadCount) &&
        nextUnreadCount >= 0
      ) {
        unreadCount.value = nextUnreadCount
      }
      return
    }
    if (
      envelope.type !== 'notice.created' ||
      !envelope.data.record ||
      typeof envelope.data.record !== 'object' ||
      Array.isArray(envelope.data.record)
    ) {
      return
    }
    const record = envelope.data.record as Record<string, any>
    const normalizedRecord = normalizeNoticeRecord(record)
    if (!Number.isSafeInteger(normalizedRecord.id) || normalizedRecord.id <= 0) {
      return
    }
    const existingIndex = records.value.findIndex((item) => item.id === normalizedRecord.id)
    if (existingIndex >= 0) {
      records.value.splice(existingIndex, 1, normalizedRecord)
    } else {
      records.value.unshift(normalizedRecord)
      if (records.value.length > size.value) {
        records.value = records.value.slice(0, size.value)
      }
      total.value += 1
    }
    const nextUnreadCount = envelope.data.unreadCount
    unreadCount.value =
      typeof nextUnreadCount === 'number' &&
      Number.isSafeInteger(nextUnreadCount) &&
      nextUnreadCount >= 0
        ? nextUnreadCount
        : unreadCount.value + 1
    latestIncomingNoticeId.value = normalizedRecord.id
  }

  const scheduleReconnect = () => {
    if (manualClose || userStore.accessMode !== 'authenticated') {
      return
    }
    if (reconnectTimer) {
      clearTimeout(reconnectTimer)
    }
    reconnectTimer = setTimeout(() => {
      connect()
    }, 3000)
  }

  const connect = () => {
    if (userStore.accessMode !== 'authenticated' || !userStore.accessToken) {
      return
    }
    if (
      socket &&
      (socket.readyState === WebSocket.OPEN || socket.readyState === WebSocket.CONNECTING)
    ) {
      return
    }
    manualClose = false
    lastSequence = 0
    const currentSocket = new WebSocket(buildNotificationWsUrl(window.location.origin), [
      NOTIFICATION_WS_PROTOCOL,
      userStore.accessToken
    ])
    socket = currentSocket
    currentSocket.onopen = () => {
      if (socket !== currentSocket) return
      if (currentSocket.protocol !== NOTIFICATION_WS_PROTOCOL) {
        manualClose = true
        currentSocket.close(1002, 'unexpected websocket protocol')
        return
      }
      connected.value = true
    }
    currentSocket.onmessage = (event) => {
      if (socket !== currentSocket) return
      try {
        const envelope = decodeNotificationWsEnvelope(JSON.parse(event.data), lastSequence)
        if (!envelope) {
          return
        }
        lastSequence = envelope.sequence
        handleWsMessage(envelope)
      } catch {
        console.warn('Failed to parse notification websocket payload')
      }
    }
    currentSocket.onclose = () => {
      if (socket !== currentSocket) return
      connected.value = false
      socket = null
      scheduleReconnect()
    }
    currentSocket.onerror = () => {
      if (socket !== currentSocket) return
      connected.value = false
    }
  }

  const disconnect = () => {
    manualClose = true
    lastSequence = 0
    connected.value = false
    if (reconnectTimer) {
      clearTimeout(reconnectTimer)
      reconnectTimer = null
    }
    if (socket) {
      socket.close()
      socket = null
    }
    resetState()
  }

  return {
    records,
    unreadCount,
    size,
    total,
    hasMore,
    loading,
    connected,
    latestIncomingNoticeId,
    loadNotices,
    markRead,
    markAllRead,
    applySignalDecision,
    connect,
    disconnect,
    resetState
  }
})

function normalizeNoticeRecord(
  raw: Api.Notifications.InAppNoticeItem | Record<string, any>
): Api.Notifications.InAppNoticeItem {
  const record = raw as Record<string, any>
  return {
    id: Number(record.id || 0),
    workflowExecutionId: record.workflowExecutionId ?? null,
    workflowExecutionNodeId: record.workflowExecutionNodeId ?? null,
    workflowDefinitionId: record.workflowDefinitionId ?? null,
    workflowDefinitionCode: String(record.workflowDefinitionCode || ''),
    workflowDefinitionName: String(record.workflowDefinitionName || ''),
    strategySignalId: record.strategySignalId ? String(record.strategySignalId) : null,
    strategySignalMode: String(record.strategySignalMode || ''),
    strategySignalStatus: String(record.strategySignalStatus || ''),
    strategySignalExpiresAt: String(record.strategySignalExpiresAt || ''),
    targetType: String(record.targetType || ''),
    targetId: record.targetId ?? null,
    targetLabel: String(record.targetLabel || ''),
    recipientId: record.recipientId ?? null,
    recipientLabel: String(record.recipientLabel || ''),
    channelType: String(record.channelType || 'in_app'),
    channelTypeLabel: String(record.channelTypeLabel || ''),
    channelDisplayName: String(record.channelDisplayName || ''),
    deliveryStatus: String(record.deliveryStatus || record.status || ''),
    deliveryStatusLabel: String(record.deliveryStatusLabel || ''),
    messageTitle: String(record.messageTitle || record.title || ''),
    messageContent: String(record.messageContent || record.content || ''),
    providerResponseText: String(record.providerResponseText || ''),
    errorMessage: String(record.errorMessage || ''),
    isRead: Boolean(record.isRead),
    readAt: String(record.readAt || ''),
    sentAt: String(record.sentAt || ''),
    createdAt: String(record.createdAt || '')
  }
}

File: backend/internal/service/notification.go
package service

import (
	"context"
	"errors"
	"sync"
	"time"

	"coinsphere/backend/internal/db"
)

type NotificationEvent struct {
	Type string `json:"type"`
	Data M      `json:"data"`
}

func (a *App) ListInAppNotifications(ctx context.Context, userID int64, page CursorPage) (M, error) {
	query := a.DB.WithContext(ctx).Model(&db.NotificationDelivery{}).
		Where("recipient_user_id = ? AND channel = ? AND status = ?", userID, "in_app", "delivered")
	var total int64
	if err := query.Count(&total).Error; err != nil {
		return nil, errors.New("count in-app notifications failed")
	}
	afterID, err := page.AfterID()
	if err != nil {
		return nil, err
	}
	if afterID > 0 {
		query = query.Where("id < ?", afterID)
	}
	var deliveries []db.NotificationDelivery
	if err := query.Order("id DESC").Limit(page.Limit + 1).Find(&deliveries).Error; err != nil {
		return nil, errors.New("list in-app notifications failed")
	}
	hasMore := len(deliveries) > page.Limit
	if hasMore {
		deliveries = deliveries[:page.Limit]
	}
	records := make([]M, len(deliveries))
	for index := range deliveries {
		records[index] = notificationRecord(deliveries[index])
	}
	lastKey := ""
	if len(deliveries) > 0 {
		lastKey = int64CursorKey(deliveries[len(deliveries)-1].ID)
	}
	result := cursorResult(records, page, lastKey, hasMore, total)
	result["unreadCount"] = a.CountUnreadInApp(ctx, userID)
	return result, nil
}

func (a *App) MarkInAppRead(ctx context.Context, userID, deliveryID int64) (int64, error) {
	now := time.Now().UTC()
	result := a.DB.WithContext(ctx).Model(&db.NotificationDelivery{}).
		Where("id = ? AND recipient_user_id = ? AND channel = ? AND status = ?", deliveryID, userID, "in_app", "delivered").
		Updates(map[string]any{"is_read": true, "read_at": now, "updated_at": now})
	if result.Error != nil {
		return 0, errors.New("mark in-app notification read failed")
	}
	if result.RowsAffected == 0 {
		return 0, ErrNotFound
	}
	return a.CountUnreadInApp(ctx, userID), nil
}

func (a *App) MarkAllInAppRead(ctx context.Context, userID int64) (M, error) {
	now := time.Now().UTC()
	result := a.DB.WithContext(ctx).Model(&db.NotificationDelivery{}).
		Where("recipient_user_id = ? AND channel = ? AND status = ? AND is_read = ?", userID, "in_app", "delivered", false).
		Updates(map[string]any{"is_read": true, "read_at": now, "updated_at": now})
	if result.Error != nil {
		return nil, errors.New("mark all in-app notifications read failed")
	}
	return M{"updatedCount": result.RowsAffected, "unreadCount": 0}, nil
}

func (a *App) CountUnreadInApp(ctx context.Context, userID int64) int64 {
	var count int64
	_ = a.DB.WithContext(ctx).Model(&db.NotificationDelivery{}).
		Where("recipient_user_id = ? AND channel = ? AND status = ? AND is_read = ?", userID, "in_app", "delivered", false).
		Count(&count).Error
	return count
}

func (a *App) PublishInAppNotification(ctx context.Context, userID, deliveryID int64) {
	var delivery db.NotificationDelivery
	if err := a.DB.WithContext(ctx).Where("id = ? AND recipient_user_id = ?", deliveryID, userID).Take(&delivery).Error; err != nil {
		return
	}
	event := NotificationEvent{Type: "notice.created", Data: M{
		"record": notificationRecord(delivery), "unreadCount": a.CountUnreadInApp(ctx, userID),
	}}
	a.notificationWatchMu.Lock()
	defer a.notificationWatchMu.Unlock()
	for watcher := range a.notificationWatches[userID] {
		select {
		case watcher <- event:
		default:
			delete(a.notificationWatches[userID], watcher)
			close(watcher)
		}
	}
	if len(a.notificationWatches[userID]) == 0 {
		delete(a.notificationWatches, userID)
	}
}

func (a *App) SubscribeNotificationEvents(ctx context.Context, userID int64) (<-chan NotificationEvent, func()) {
	watcher := make(chan NotificationEvent, 64)
	a.notificationWatchMu.Lock()
	watcher <- NotificationEvent{Type: "notice.unread", Data: M{"unreadCount": a.CountUnreadInApp(ctx, userID)}}
	if a.notificationWatches[userID] == nil {
		a.notificationWatches[userID] = map[chan NotificationEvent]struct{}{}
	}
	a.notificationWatches[userID][watcher] = struct{}{}
	a.notificationWatchMu.Unlock()
	var once sync.Once
	return watcher, func() {
		once.Do(func() {
			a.notificationWatchMu.Lock()
			if _, exists := a.notificationWatches[userID][watcher]; exists {
				delete(a.notificationWatches[userID], watcher)
				close(watcher)
			}
			if len(a.notificationWatches[userID]) == 0 {
				delete(a.notificationWatches, userID)
			}
			a.notificationWatchMu.Unlock()
		})
	}
}

func (a *App) CloseNotificationEvents() {
	a.notificationWatchMu.Lock()
	defer a.notificationWatchMu.Unlock()
	for _, watchers := range a.notificationWatches {
		for watcher := range watchers {
			close(watcher)
		}
	}
	a.notificationWatches = map[int64]map[chan NotificationEvent]struct{}{}
}

func notificationRecord(delivery db.NotificationDelivery) M {
	return M{
		"id": delivery.ID, "workflowExecutionId": nil, "workflowExecutionNodeId": nil,
		"workflowDefinitionId": delivery.WorkflowID, "workflowDefinitionCode": "", "workflowDefinitionName": "",
		"strategySignalId": nil, "strategySignalMode": "", "strategySignalStatus": "", "strategySignalExpiresAt": "",
		"targetType": "user", "targetId": delivery.RecipientUserID, "targetLabel": "",
		"recipientId": delivery.RecipientUserID, "recipientLabel": "",
		"channelType": delivery.Channel, "channelTypeLabel": "站内通知", "channelDisplayName": "站内通知",
		"deliveryStatus": delivery.Status, "deliveryStatusLabel": "已送达",
		"messageTitle": delivery.Title, "messageContent": delivery.Message,
		"providerResponseText": "", "errorMessage": "", "isRead": delivery.IsRead,
		"readAt": fmtTime(delivery.ReadAt), "sentAt": fmtTime(delivery.DeliveredAt),
		"createdAt": delivery.CreatedAt.UTC().Format(time.RFC3339Nano),
	}
}

func fmtTime(value *time.Time) string {
	if value == nil {
		return ""
	}
	return value.UTC().Format(time.RFC3339Nano)
}
